src/adapter/src/coord/ddl.rs

```diff
@@ -709,16 +665,20 @@ impl<S: Append + 'static> Coordinator<S> {
     /// Validate a specific type of resource limit and return an error if that limit is exceeded.
     fn validate_resource_limit<F>(
         &self,
-        current_amount: i32,
-        new_instances: i32,
+        current_amount: usize,
+        new_instances: u32,
         resource_limit: F,
         resource_type: &str,
     ) -> Result<(), AdapterError>
     where
-        F: Fn(&SystemVars) -> i32,
+        F: Fn(&SystemVars) -> u32,
     {
         let limit = resource_limit(self.catalog.state().system_config());
-        if new_instances > 0 && current_amount + new_instances > limit {
+        if current_amount > usize::cast_from(u32::MAX)
+            || (new_instances > 0
+                && u32::try_from(current_amount).expect("guaranteed to fit") + new_instances
+                    > limit)
+        {
```

Comment:
Not that it really matters, but you might be able to avoid the expect with something like this?

```
let exceeds_limit = match u32::try_from(current_amount) {
    Ok(current_amount) => new_instances > 0 && current_amount + new_instances > limit,
    Err(_) => true,
}
```

Also, is this robust against overflows? It looks to me like if `current_amount` is `u32::MAX` and limit is `u32::MAX` the code (both yours and mine) will wrap around and erroneously allow the creation of the resource. Unlikely to matter in practice, because we likely won't have limits that large, but now that I've seen it I can't unsee it.

Reply:
Yeah, you're right. I should be using `checked_add` here. I'll fix that.